woocommerce/woocommerce-android
# Decouple reviewmoderation flow between ReviewList and ReviewDetail Fr…

Decouple review moderation flow between Reviewlist and ReviewDetail fragment 



Closes: #5647 


### Depends on : [this PR](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2271)

### Description
This is not a bug per se but a redesign for communication of the review moderation flow between from ReviewList Fragment 
and ReviewDetailFragment . This redesign makes it easy for ReviewDetail Fragment to be re used from another fragment for moderation. 

### The Goals for this was 

1. Remove the usage of EventBus from the communication between the screens.
2. Make ReviewDetailFragment independent from having ReviewListFragment on the backstack.
3. Keep the logic of async moderation: show a Snackbar on the parent screen with undo button.
4. Nice to have: migrate newUpdateProductReviewStatusAction to a suspendable function.

### Design
 This design is governed by ReviewModeration Interface which
 has 3 parts implemented by 3 classes 

  The first is the ReviewModerationHandler
  This class contains all the business logic for Review Moderation
  It is responsible for
  a) Receiving the moderation event from DetailReviewFragment
  b) Updating the ProductStore with review status and share the result
  c) Decide what needs to be done to the UI at various stage of moderation
  e) It primarily publishes events via shared flow
 
  The next is the class implementing relay interface
  A base class implementation is done as ViewModel
  This class receives events from the Handler via a collector
  and "relays" them to the View as Live Data events .
 
  Finally  we come to the View
  This class implements all the UI related functionality for showing the intermediate  moderation UI
  (like showing undo snackbar and refresh loader) and updating
  the Reviews list as per the events received from handler via the Relay Class

### Testing instructions


1. Go to Reviews Section of the app (ensure enough review is there)
2. Go to review detail choose any of the moderation action
3. See the changes reflected in the RviewList UI (There is no change in functionality in this implementation) 

### Images/gif


### Note 
This needs the FluxC hash of the dependent PR to compile.
